lib/oxttl/src/toolkit/chunker.rs

```diff
@@ -0,0 +1,107 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use crate::TurtleParser;
+
+const EOT_CHAR: u8 = b'.';
+
+// Given a number of desired chunks, corresponding to threads find offsets that break the file into chunks that can be read in parallel.
+// Parser should not be reused, hence it is passed by value.
+#[allow(clippy::needless_pass_by_value)]
+pub fn get_turtle_file_chunks(
+    bytes: &[u8],
+    n_chunks: usize,
+    parser: TurtleParser,
+) -> Vec<(usize, usize)> {
+    let mut last_pos = 0;
+    let total_len = bytes.len();
+    let chunk_size = total_len / n_chunks;
+    let mut offsets = Vec::with_capacity(n_chunks);
+    for _ in 0..n_chunks {
+        let search_pos = last_pos + chunk_size;
+
+        if search_pos >= bytes.len() {
+            break;
+        }
+
+        let end_pos = match next_terminating_period(parser.clone(), &bytes[search_pos..]) {
+            Some(pos) => search_pos + pos,
+            None => {
+                return vec![(0, total_len)];
+            }
+        };
+        offsets.push((last_pos, end_pos));
+        last_pos = end_pos;
+    }
+    offsets.push((last_pos, total_len));
+    offsets
+}
+
+// Heuristically, we assume that a period is terminating (a triple) if we can start immediately after it and parse three triples.
+// Parser should not be reused, hence it is passed by value.
+// If no such period can be found, looking at 1000 consecutive periods, we give up.
+// Important to keep this number this high, as some TTL files can have a lot of periods.
+#[allow(clippy::needless_pass_by_value)]
+fn next_terminating_period(parser: TurtleParser, mut input: &[u8]) -> Option<usize> {
+    fn accept(parser: TurtleParser, input: &[u8]) -> bool {
+        let mut f = parser.parse_slice(input);
+        for _ in 0..3 {
+            if let Some(r) = f.next() {
+                if r.is_err() {
+                    return false;
+                }
+            } else {
+                return false;
+            }
+        }
+        true
+    }
+
+    let mut rejected_periods = 0_u16;
+
+    let mut total_pos = 0;
+    if input.is_empty() {
+        return None;
+    }
+    loop {
+        if rejected_periods >= 1_000 {
+            return None;
+        }
```

Comment:
nit: a `for _ in 0..1_000` is slightly shorter

Reply:
Fixed